File: src/main/java/com/freya02/botcommands/api/parameters/ParameterResolver.java
package com.freya02.botcommands.api.parameters;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.core.entities.UserUnion;
import com.freya02.botcommands.api.core.service.annotations.InterfacedService;
import com.freya02.botcommands.api.core.utils.ReflectionUtils;
import kotlin.reflect.KClass;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Message;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel;
import net.dv8tion.jda.api.entities.emoji.Emoji;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

/**
 * Base class for parameter resolvers used in regex commands, application commands and buttons callbacks
 * <p>
 * Parameters supported by default:
 * <ul>
 *     <li>{@link String}</li>
 *
 *     <li>boolean</li>
 *     <li>long</li>
 *     <li>double</li>
 *
 *     <li>{@link Emoji}</li>
 *
 *     <li>{@link Role}</li>
 *     <li>{@link User}</li>
 *     <li>{@link Member}</li>
 *     <li>{@link UserUnion}</li>
 *     <li>{@link GuildChannel all guild channels subtypes (in theory)}</li>
 *     <li>{@link Message} (only message context commands)</li>
 * </ul>
 *
 * You can also check loaded parameter resolvers in the logs, on the "trace" level
 *
 * @see RegexParameterResolver
 * @see QuotableRegexParameterResolver
 * @see ComponentParameterResolver
 * @see SlashParameterResolver
 * @see MessageContextParameterResolver
 * @see UserContextParameterResolver
 * @see ICustomResolver
 */
@SuppressWarnings("unused") //T is used for the inheritance constraint
@InterfacedService(acceptMultiple = true)
public abstract class ParameterResolver<T extends ParameterResolver<T, R>, R> {
	protected final Logger LOGGER = Logging.getLogger(this);

	private final KClass<R> jvmErasure;

	/**
	 * Constructs a new parameter resolver
	 *
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolver(@NotNull Class<R> clazz) {
		this.jvmErasure = ReflectionUtils.toKotlin(clazz);
	}

	/**
	 * Constructs a new parameter resolver
	 *
	 * @param clazz Class of the parameter being resolved
	 */
	public ParameterResolver(@NotNull KClass<R> clazz) {
		this.jvmErasure = clazz;
	}

	public KClass<R> getJvmErasure() {
		return jvmErasure;
	}
}

File: src/main/kotlin/io/github/freya022/botcommands/api/parameters/ParameterResolver.kt
package io.github.freya022.botcommands.api.parameters

import io.github.freya022.botcommands.api.commands.application.slash.annotations.ChannelTypes
import io.github.freya022.botcommands.api.commands.application.slash.annotations.MentionsString
import io.github.freya022.botcommands.api.core.entities.InputUser
import io.github.freya022.botcommands.api.core.service.annotations.BService
import io.github.freya022.botcommands.api.core.service.annotations.InterfacedService
import io.github.freya022.botcommands.api.parameters.resolvers.*
import net.dv8tion.jda.api.entities.*
import net.dv8tion.jda.api.entities.Message.Attachment
import net.dv8tion.jda.api.entities.channel.middleman.GuildChannel
import net.dv8tion.jda.api.entities.emoji.Emoji

/**
 * Base class for parameter resolvers used in text commands, application commands, and component callbacks.
 *
 * You need to extend [ClassParameterResolver] or [TypedParameterResolver] instead.
 *
 * ### Default parameter resolvers
 *
 * |                                                               | Text              | Slash           | Message context    | User context    | Components | Modals | Component timeout |
 * |---------------------------------------------------------------|-------------------|-----------------|--------------------|-----------------|------------|--------|-------------------|
 * | [String]                                                      | ✓ (can be quoted) | ✓               |                    |                 | ✓          | ✓      | ✓
 * | [Boolean]                                                     | ✓                 | ✓               |                    |                 | ✓          |        | ✓
 * | [Int]                                                         | ✓                 | ✓               |                    |                 | ✓          |        | ✓
 * | [Long]                                                        | ✓                 | ✓               |                    |                 | ✓          |        | ✓
 * | [Double]                                                      | ✓                 | ✓               |                    |                 | ✓          |        | ✓
 * | [Emoji]                                                       | ✓                 | ✓               |                    |                 | ✓          |        | ✓
 * | [IMentionable]                                                | ✓ (only mentions) | ✓               |                    |                 |            |        |
 * | [List] of mentionable (see [@MentionsString][MentionsString]) |                   | ✓               |                    |                 |            |        |
 * | [Role]                                                        | ✓                 | ✓               |                    |                 | ✓          |        |
 * | [User]                                                        | ✓                 | ✓               |                    | ✓ (target user) | ✓          |        |
 * | [Member]                                                      | ✓                 | ✓               |                    | ✓ (target user) | ✓          |        |
 * | [InputUser]                                                   | ✓                 | ✓               |                    | ✓ (target user) | ✓          |        |
 * | [GuildChannel] subtypes<sup>1</sup>                           | ✓                 | ✓               |                    |                 | ✓          |        |
 * | [Guild]                                                       | ✓                 | ✓ (as a String) |                    |                 | ✓          |        |
 * | [Message]                                                     |                   |                 | ✓ (target message) |                 |            |        |
 * | [Attachment]                                                  |                   | ✓               |                    |                 |            |        |
 *
 * 1. The channel types are set automatically depending on the type,
 * but a broader channel type can be used
 * and restricted to multiple concrete types by using [@ChannelTypes][ChannelTypes].
 *
 * Parameter resolvers for services exist by default, and follow the rules described in [@BService][BService].
 *
 * You can also check loaded parameter resolvers in the logs on the `trace` level.
 *
 * @param T Type of the implementation
 * @param R Type of the returned resolved objects
 *
 * @see ClassParameterResolver
 *
 * @see ParameterResolverFactory
 *
 * @see TextParameterResolver
 * @see QuotableTextParameterResolver
 * @see ComponentParameterResolver
 * @see SlashParameterResolver
 * @see MessageContextParameterResolver
 * @see UserContextParameterResolver
 * @see TimeoutParameterResolver
 * @see ICustomResolver
 *
 * @see Resolvers
 */
@InterfacedService(acceptMultiple = true)
sealed class ParameterResolver<T : ParameterResolver<T, R>, R : Any> : IParameterResolver<T>
